src/main/scala/piconot/external/parser/Parser.scala

```diff
@@ -0,0 +1,60 @@
+package piconot.external
+
+import scala.util.parsing.combinator._
+import PiconotTypes._
+
+
+/** A recursive descent parser for the Piconot language. See design.md for
+ *  the grammar.
+ */
+object PiconotParser extends JavaTokenParsers with PackratParsers with RegexParsers {
+
+  // parsing interface
+  def apply(s: String): ParseResult[Seq[State]] = {
+    val noComments = s.replaceAll("#.*(\n|$)", " ")
+    parseAll(program, noComments)
+  }
```

Comment:
Your parser looks very elegant! My one comment would be that it passes through twice because of the way you implemented stripping the comments, but I think it makes for a clean solution, so it's not a big deal.